### PR TITLE
Correctly throw IndexOutOfBoundsException when dst.remaining() is too…

### PR DESCRIPTION
**Pull request by normanmaurer**
… big.

Motivation:

In some ByteBuf implementations we not correctly implement getBytes(index, ByteBuffer).

Modifications:

Correct code to do what is defined in the javadocs and adding test.

Result:

Implementation works as descripted.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @trustin PTAL

File: buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java [hunk 1]
```diff
@@ -341,19 +341,18 @@ public ByteBuf getBytes(int index, ByteBuffer dst) {
     }
 
     private void getBytes(int index, ByteBuffer dst, boolean internal) {
-        checkIndex(index);
+        checkIndex(index, dst.remaining());
```
Comment (netkins) [inline buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java]: CRITICAL Nullcheck of dst at line 345 of value previously dereferenced in io.netty.buffer.UnpooledDirectByteBuf.getBytes(int, ByteBuffer, boolean) [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE)

**Comment by mosesn** [comment]
ffti: s/descripted/described in the commit message.

**Comment by normanmaurer** [comment]
@trustin thanks... let us just hope not too many people depend on the incorrect behaviour 👍

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (ed931b19bbef9e4080dbf5ee0f3c89a0e266f530) and 4.1 (97bf3c0a9bcd6e3f7f9ad5465779242c0edde390)

**Pull request closed without merging**